processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+    for (int oldBucket = 0; oldBucket < buckets; oldBucket++) {
+      if (isUsed(oldBucket)) {
+        final ByteBuffer entryBuffer = tableBuffer.duplicate();
+        entryBuffer.position(oldBucket * bucketSize);
+        entryBuffer.limit((oldBucket + 1) * bucketSize);
+
+        final ByteBuffer keyBuffer = entryBuffer.duplicate();
+        keyBuffer.position(entryBuffer.position() + HASH_SIZE);
+        keyBuffer.limit(keyBuffer.position() + keySize);
+
+        final int keyHash = entryBuffer.getInt(entryBuffer.position()) & 0x7fffffff;
+
+        final int newBucket = findBucket(newTableBuffer, newBuckets, bucketSize, true, keyBuffer, keySize, keyHash);
+        if (newBucket < 0) {
+          throw new ISE("WTF?! Couldn't find a bucket while resizing?!");
+        }
+        newTableBuffer.position(newBucket * bucketSize);
+        newTableBuffer.put(entryBuffer);
+
+        buffer.putInt(tableArenaSize + newSize * Ints.BYTES, newBucket * bucketSize);
+        newSize++;
+      }
+    }
+
+    buckets = newBuckets;
+    maxSize = newMaxSize;
+    tableBuffer = newTableBuffer;
+    tableStart = newTableStart;
+
+    if (size != newSize) {
+      throw new ISE("WTF?! size[%,d] != newSize[%,d] after resizing?!", size, maxSize);
+    }
+  }
+
+  private static int maxSizeForBuckets(int buckets)
+  {
+    return Math.max(1, (int) (buckets * MAX_LOAD_FACTOR));
+  }
+
+  /**
+   * Finds the bucket into which we should insert a key.
+   *
+   * @param keyBuffer key, must have exactly keySize bytes remaining
+   *
+   * @return bucket index for this key, or -1 if no bucket is available due to being full
+   */
+  private static int findBucket(
+      final ByteBuffer tableBuffer,
+      final int buckets,
+      final int bucketSize,
+      final boolean allowNewBucket,
+      final ByteBuffer keyBuffer,
+      final int keySize,
+      final int keyHash
+  )
+  {
+    Preconditions.checkArgument(
+        keySize == keyBuffer.remaining(),
+        "keyBuffer.remaining[%s] != keySize[%s]",
+        keyBuffer.remaining(),
+        keySize
+    );
+
+    final int startBucket = Math.abs(keyHash % buckets);
+    int bucket = startBucket;
+
+outer:
+    while (true) {
+      final int bucketOffset = bucket * bucketSize;
+
+      if ((tableBuffer.get(bucketOffset) & 0x80) == 0) {
+        // Found unused bucket before finding our key
+        return allowNewBucket ? bucket : -1;
+      }
+
+      for (int i = bucketOffset + HASH_SIZE, j = keyBuffer.position(); j < keyBuffer.position() + keySize; i++, j++) {
+        if (tableBuffer.get(i) != keyBuffer.get(j)) {
+          bucket += 1;
+          if (bucket == buckets) {
+            bucket = 0;
+          }
+
+          if (bucket == startBucket) {
+            // Came back around to the start without finding a free slot, that was a long trip!
+            // Should never happen unless buckets == maxSize.
+            return -1;
+          }
+
+          continue outer;
```

Comment:
hmmm, it looks like we will need another variable if we wanted to avoid the label. i'm not used to seeing labels used much in java so wanted to avoid same.... that said it might just be personal preference.